### ISSUE TITLE
Improve AND filter operator

### ISSUE THREAD

**Issue by Jackie-Jiang** [apache/pinot#9839]
Related issues: #9402, #8837

There are some improvements we want to do for AND filter operator:
- Avoid scanning the docs from the low selectivity filter when applying AND
- Skip using indexes when the selected docs are already very few. Using scan instead
- When applying bitmap to a scanned operator, push down the limit to avoid full scan (or evaluate lazily)

**Comment by ashishkf** [comment]
Can you please see if docs can be iterated per the "order by" clause if the order by column also happens to be in the sorted column? This is useful if say, segments are sorted by the timestamp column but the query has "order by timestamp desc". In that case, in addition to limit pushdown applyAnd could passed a flag - ascOrDesc - to use either the forward or reverse bitmap iterator.

**Comment by Jackie-Jiang** [comment]
@ashishkf Good point! There are actually some discussion about providing reversed doc iterator in #8979. We should consider adding a method to return the reversed iterator, and evaluate `applyAnd` lazily.
